File: packages/axon-core-ts/src/coa.ts
import type { AttachmentRef } from "./lot.js";

export type MeasurementResult = "pass" | "fail" | "review" | "informational";

/**
 * One row from a Certificate of Analysis. Values can be numeric or string
 * (e.g. "<1 ppm" appears verbatim in supplier CoAs and must round-trip).
 */
export interface CoAMeasurement {
  parameter: string;
  value: number | string;
  unit: string;
  spec?: { min?: number; max?: number; lt?: number; gt?: number };
  result: MeasurementResult;
  notes?: string;
}

export interface CoA {
  id: string;
  lotId: string;
  supplierId: string;
  issuedAt: string; // ISO-8601
  validatedBy?: string;
  measurements: CoAMeasurement[];
  attachments?: AttachmentRef[];
  /** Overall pass/fail rolled up from measurements. */
  overall: MeasurementResult;
}

export function rollUp(measurements: CoAMeasurement[]): MeasurementResult {
  if (measurements.some((m) => m.result === "fail")) return "fail";
  if (measurements.some((m) => m.result === "review")) return "review";
  return "pass";
}

/** Apply a spec to a numeric value and return the verdict. */
export function evaluate(value: number, spec: NonNullable<CoAMeasurement["spec"]>): MeasurementResult {
  if (spec.min !== undefined && value < spec.min) return "fail";
  if (spec.max !== undefined && value > spec.max) return "fail";
  if (spec.lt !== undefined && !(value < spec.lt)) return "fail";
  if (spec.gt !== undefined && !(value > spec.gt)) return "fail";
  return "pass";
}

File: packages/axon-core-py/src/axon_core/coa.py
from __future__ import annotations
from typing import Literal, Optional, Union
from pydantic import BaseModel, Field
from .lot import AttachmentRef

MeasurementResult = Literal["pass", "fail", "review", "informational"]


class _Spec(BaseModel):
    min: Optional[float] = None
    max: Optional[float] = None
    lt: Optional[float] = None
    gt: Optional[float] = None


class CoAMeasurement(BaseModel):
    parameter: str
    value: Union[float, str]
    unit: str
    spec: Optional[_Spec] = None
    result: MeasurementResult
    notes: Optional[str] = None


class CoA(BaseModel):
    id: str
    lotId: str
    supplierId: str
    issuedAt: str
    validatedBy: Optional[str] = None
    measurements: list[CoAMeasurement] = Field(default_factory=list)
    attachments: Optional[list[AttachmentRef]] = None
    overall: MeasurementResult


def roll_up(measurements: list[CoAMeasurement]) -> MeasurementResult:
    if any(m.result == "fail" for m in measurements):
        return "fail"
    if any(m.result == "review" for m in measurements):
        return "review"
    return "pass"


def evaluate(value: float, spec: _Spec) -> MeasurementResult:
    if spec.min is not None and value < spec.min:
        return "fail"
    if spec.max is not None and value > spec.max:
        return "fail"
    if spec.lt is not None and not (value < spec.lt):
        return "fail"
    if spec.gt is not None and not (value > spec.gt):
        return "fail"
    return "pass"
